File: src/main/java/kanti/sl/ValueConverterImpl.java
package kanti.sl;

import kanti.sl.arguments.TypeChecker;
import org.jetbrains.annotations.NotNull;

public class ValueConverterImpl implements ValueConverter {

	private final boolean supportPrimitive;

	public ValueConverterImpl(boolean supportPrimitive) {
		this.supportPrimitive = supportPrimitive;
	}

	@NotNull
	@Override
	public String convert(@NotNull Object value) {
		if (!TypeChecker.check(value)) {
			Class<?> klazz = value.getClass();
			throw new IllegalArgumentException("Unsupported type=" + klazz.getSimpleName());
		}
		return value.toString();
	}

	@NotNull
	@Override
	public Object convert(@NotNull String line) {
		if (supportPrimitive) {
			return convertToValueWithPrimitive(line);
		} else {
			return line;
		}
	}

	@NotNull
	private Object convertToValueWithPrimitive(@NotNull String line) {
		if (line.equals("true") || line.equals("false")) {
			return Boolean.parseBoolean(line);
		} else if (line.contains(".")) {
			return Double.parseDouble(line);
		} else {
			try {
				return Integer.parseInt(line);
			} catch (NumberFormatException ex) {
				return line;
			}
		}
	}

	static class Builder implements ValueConverter.Builder {

		private boolean supportPrimitive = true;

		@NotNull
		@Override
		public ValueConverter.Builder supportPrimitive(boolean support) {
			supportPrimitive = support;
			return this;
		}

		@NotNull
		@Override
		public ValueConverter build() {
			return new ValueConverterImpl(supportPrimitive);
		}

	}

}

File: src/main/kotlin/kanti/sl/ValueConverterImpl.kt
package kanti.sl

import kanti.sl.arguments.TypeChecker

class ValueConverterImpl(
	private val supportPrimitive: Boolean
) : ValueConverter {

	override fun convert(value: Any): String {
		require(TypeChecker.check(value)) {
			"Unsupported type: ${value.javaClass.simpleName}"
		}
		return value.toString()
	}

	override fun convert(line: String): Any {
		return if (supportPrimitive) {
			convertToValueWithPrimitive(line)
		} else {
			return line
		}
	}

	private fun convertToValueWithPrimitive(line: String): Any {
		if (line == "true" || line == "false")
			return line.toBoolean()
		if (line.contains('.'))
			return line.toDouble()
		return line.toIntOrNull() ?: line
	}

	class Builder : ValueConverter.Builder {

		private var supportPrimitive: Boolean = true

		override fun supportPrimitive(support: Boolean): ValueConverter.Builder {
			supportPrimitive = support
			return this
		}

		override fun build(): ValueConverter {
			return ValueConverterImpl(
				supportPrimitive = supportPrimitive
			)
		}
	}

}